inspektor-gadget/inspektor-gadget
# run: Only enable relevant operators according to the gadget being run

This PR handles the issues described in #2167. 

### How to test

Considering the following simple gadget that doesn't provide any mount ns related info:

```c
// program.bpf.c

// Kernel types definitions
// Check https://blog.aquasec.com/vmlinux.h-ebpf-programs for more details
#include <vmlinux.h>

// eBPF helpers signatures
// Check https://man7.org/linux/man-pages/man7/bpf-helpers.7.html to learn
// more about different available helpers
#include <bpf/bpf_helpers.h>

// Inspektor Gadget macros
#include <gadget/macros.h>

struct event {
	__u32 pid;
};

// we need to "use" this struct for the compiler not to remove it in an optimization pass
const struct event *unusedevent __attribute__((unused));

struct {
	__uint(type, BPF_MAP_TYPE_PERF_EVENT_ARRAY);
	__uint(key_size, sizeof(u32));
	__type(value, struct event);
} events SEC(".maps");

GADGET_TRACE_MAP(events);

SEC("tracepoint/syscalls/sys_enter_openat")
int enter_openat(struct trace_event_raw_sys_enter *ctx)
{
	struct event event = {};

	event.pid = bpf_get_current_pid_tgid() >> 32;

	bpf_perf_event_output(ctx, &events, BPF_F_CURRENT_CPU,
			      &event, sizeof(event));

	return 0;
}

char LICENSE[] SEC("license") = "GPL";
```

and its metadata file

```yaml
name: 'TODO: Fill the gadget name'
description: 'TODO: Fill the gadget description'
tracers:
  events:
    mapName: events
    structName: event
structs:
  event:
    fields:
    - name: pid
      description: 'TODO: Fill field description'
      attributes:
        width: 10
        alignment: left
        ellipsis: end
```

Build and tag is as `mygadget`

```bash
$ sudo -E ig image build .  -t mygadget
```

### Before this PR 

```bash
$ sudo -E ig run mygadget -h
INFO[0000] Experimental features enabled                
Run a containerized gadget
...
```

The gadget provides some filtering options that don't do anything
```bash 
   -c, --containername string   Show only data from containers with that name (default ""
...
      --host                   Show data from both the host and containers
```

And a lot of possible columns that aren't really available

```bash
                                Available columns:
                                     k8s.container
                                     k8s.hostnetwork
                                     k8s.namespace
                                     k8s.node
                                     k8s.pod
                                     mntns                                     
                                     runtime.containerId
                                     runtime.containerImageDigest
                                     runtime.containerImageName
                                     runtime.containerName
                                     runtime.runtimeName

```

```bash
$ sudo -E ig run mygadget 
INFO[0000] Experimental features enabled                
RUNTIME.CONTAINERNAME                                                                                                  PID                                    
                                                                                                                       8626                                   
                                                                                                                       3516                                   
                                                                                                                       1140                                   
                                                                                                                       1140                                   
                                                                                                                       1140                                   
                                                                                                                       1140                                   
                                                                                                                       1140                                   
                                                                                                                       1140                                   
                                                                                                                       1140                                   
                                                                                                                       1140    
```

A mount ns set map that is not used is created:

```bash
$ sudo bpftool map list | grep mntnsset -A 2
1996: hash  name mntnsset_21a9da  flags 0x0
	key 8B  value 4B  max_entries 1024  memlock 16384B
	pids ig(438531)
```

### After this PR 

Filtering options and columns aren't shown anymore:
(we still need to fix the timestamp one)

```bash
                                Available columns:
                                  pid
                                  timestamp
```

```bash
$ sudo -E ig run mygadget 
INFO[0000] Experimental features enabled                
PID                                                                                                                                                           
3516                                                                                                                                                          
2915                                                                                                                                                          
2915                                                                                                                                                          
2915                                                                                                                                                          
2915                                                                                                                                                          
2915                                                                                                                                                          
2915                                                                                                                                                          
442243 
```

And the mount ns set map is not created:

```bash
$ sudo bpftool map list | grep mntnsset -A 2
```
